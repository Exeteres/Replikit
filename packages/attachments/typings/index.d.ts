export * from "./entities/attachment";

export * from "./extensions/storage";
