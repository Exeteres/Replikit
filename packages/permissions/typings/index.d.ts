export * from "./userRoleMap";
export * from "./memberRoleMap";
export * from "./userPermissionMap";
export * from "./memberPermissionMap";
export * from "./typeMap";
export * from "./roleInfo";
export * from "./extensions/storage";
