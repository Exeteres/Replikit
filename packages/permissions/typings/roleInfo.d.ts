import {
    TypeName,
    RoleName,
    PermissionName
} from "@replikit/permissions/typings";

export interface RoleInfo<T extends TypeName = TypeName> {
    type: TypeName;
    name: RoleName<T>;
    permissions: PermissionName<T>[];
    fallbackRoles: RoleInfo[];
}
