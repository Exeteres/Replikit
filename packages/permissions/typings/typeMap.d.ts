import {
    UserPermissionMap,
    MemberPermissionMap,
    UserRoleMap,
    MemberRoleMap
} from "@replikit/permissions/typings";

export interface TypeMap {
    user: [UserRoleMap, UserPermissionMap];
    member: [MemberRoleMap, MemberPermissionMap];
}

export type TypeName = keyof TypeMap;

type _RoleName<T extends TypeName> = keyof TypeMap[T][0];

export type RoleName<
    T extends TypeName = TypeName
> = /**************/ _RoleName<T> extends never ? unknown : _RoleName<T>;

type _PermissionName<T extends TypeName> = keyof TypeMap[T][1];

export type PermissionName<
    T extends TypeName = TypeName
> = /****/ _PermissionName<T> extends never ? unknown : _PermissionName<T>;

export type UserPermissionName = PermissionName<"user">;
export type MemberPermissionName = PermissionName<"member">;
