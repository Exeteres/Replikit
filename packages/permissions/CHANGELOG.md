# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [0.9.0](https://github.com/Exeteres/Replikit/compare/v0.8.0...v0.9.0) (2020-06-28)

**Note:** Version bump only for package @replikit/permissions





# [0.8.0](https://github.com/Exeteres/Replikit/compare/v0.7.2...v0.8.0) (2020-06-27)


### Bug Fixes

* **permissions:** Add UserRoleName and MemberRoleName aliases ([b0c8199](https://github.com/Exeteres/Replikit/commit/b0c81999045660cd4ac976376c178d97ec6fb947))





## [0.7.2](https://github.com/Exeteres/Replikit/compare/v0.7.1...v0.7.2) (2020-06-23)

**Note:** Version bump only for package @replikit/permissions





## [0.7.1](https://github.com/Exeteres/Replikit/compare/v0.7.0...v0.7.1) (2020-06-23)

**Note:** Version bump only for package @replikit/permissions





# [0.7.0](https://github.com/Exeteres/Replikit/compare/v0.6.0...v0.7.0) (2020-06-23)


### Bug Fixes

* **permissions:** Fix missing storage defaults ([6e6ca9d](https://github.com/Exeteres/Replikit/commit/6e6ca9dd77267a198ebd66a019e601c508dabbcb))





# [0.6.0](https://github.com/Exeteres/Replikit/compare/v0.5.3...v0.6.0) (2020-06-21)


### Features

* **permissions:** Add api for getting permission and role names ([e7cba6c](https://github.com/Exeteres/Replikit/commit/e7cba6c1d1ed6cf76916fff1afc577046c3c0d55))
* **storage:** Add more useful methods ([63e5125](https://github.com/Exeteres/Replikit/commit/63e512520c0d1092ac175af3c9be14b05a4051ae))





## [0.5.3](https://github.com/Exeteres/Replikit/compare/v0.5.2...v0.5.3) (2020-06-20)

**Note:** Version bump only for package @replikit/permissions





## [0.5.2](https://github.com/Exeteres/Replikit/compare/v0.5.1...v0.5.2) (2020-06-20)

**Note:** Version bump only for package @replikit/permissions





# 0.5.0 (2020-06-20)

**Note:** Version bump only for package @replikit/permissions






# [0.4.0](https://github.com/Exeteres/Replikit/compare/v0.3.0...v0.4.0) (2020-06-19)

**Note:** Version bump only for package @replikit/permissions





# [0.3.0](https://github.com/Exeteres/Replikit/compare/v0.2.0...v0.3.0) (2020-06-14)

**Note:** Version bump only for package @replikit/permissions





# [0.2.0](https://github.com/Exeteres/Replikit/compare/v0.1.0...v0.2.0) (2020-06-13)

**Note:** Version bump only for package @replikit/permissions
