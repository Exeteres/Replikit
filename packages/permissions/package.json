{
    "name": "@replikit/permissions",
    "version": "0.7.0",
    "license": "MIT",
    "main": "dist/index.js",
    "typings": "dist/index.d.ts",
    "files": [
        "dist",
        "typings"
    ],
    "scripts": {
        "build": "ttsc -p tsconfig.build.json"
    },
    "dependencies": {
        "@replikit/core": "^0.7.0",
        "@replikit/storage": "^0.7.0"
    }
}
