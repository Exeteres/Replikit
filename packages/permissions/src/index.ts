/// <reference path="../typings/index.d.ts" />

export * from "./models/enum";
export * from "./models/entityType";
export * from "./models/permission";
export * from "./models/role";
export * from "./models/memberPermission";
export * from "./models/memberRole";
export * from "./models/userPermission";
export * from "./models/userRole";

export * from "./errors";
export * from "./permissionsLocale";
export * from "./permissionStorage";
export * from "./hasPermissions";

import "./locales/en";
import "./locales/ru";

export * from "./startup";
