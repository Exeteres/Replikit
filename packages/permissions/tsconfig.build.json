{
    "extends": "../../tsconfig.build.json",
    "compilerOptions": {
        "outDir": "dist",
        "paths": {
            "@replikit/permissions": ["permissions/src"]
        }
    },
    "include": ["src"]
}
