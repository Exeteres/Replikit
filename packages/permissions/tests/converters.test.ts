import {
    MemberPermission,
    UserPermission,
    UserRole,
    MemberRole,
    registerPermissionsConverters,
    permissionStorage
} from "@replikit/permissions";
import "@replikit/permissions/tests";
import { createTestManager, TestManagerSuite } from "@replikit/test-utils";
import { Constructor } from "@replikit/core/typings";

function createConverterTestManager(): TestManagerSuite {
    const suite = createTestManager();
    registerPermissionsConverters(suite.converter, permissionStorage);
    return suite;
}

describe("converters", () => {
    it.each([
        ["user permission", UserPermission],
        ["member permission", MemberPermission],
        ["user role", UserRole],
        ["member role", MemberRole]
    ])("should resolve %s parameter", async (_, type) => {
        const { command, testManager } = createConverterTestManager();
        command("test")
            .required("param", type as Constructor)
            .handler(context => expect(context.params.param).toMatchSnapshot())
            .register();
        await testManager.processCommand("/test Test1");
        expect.assertions(1);
    });

    it.each([
        ["user permission", UserPermission],
        ["member permission", MemberPermission],
        ["user role", UserRole],
        ["member role", MemberRole]
    ])("should throw an error if %s parameter is invalid", async (_, type) => {
        const { command, testManager } = createConverterTestManager();
        command("test")
            .required("param", type as Constructor)
            .handler(context => {
                throw void context;
            })
            .register();
        await testManager.processCommand("/test Test10");
        expect.assertions(1);
    });
});
