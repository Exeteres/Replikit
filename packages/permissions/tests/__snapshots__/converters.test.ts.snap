// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`converters should resolve member permission parameter 1`] = `
TestMemberPermission {
  "id": "test0",
  "index": 0,
  "name": "Test1",
  "type": EntityType {
    "name": "Member",
  },
}
`;

exports[`converters should resolve member role parameter 1`] = `
TestMemberRole {
  "fallbackRoles": Array [],
  "id": "test0",
  "index": 0,
  "name": "Test1",
  "permissions": Array [],
  "type": EntityType {
    "name": "Member",
  },
}
`;

exports[`converters should resolve user permission parameter 1`] = `
TestUserPermission {
  "id": "test0",
  "index": 0,
  "name": "Test1",
  "type": EntityType {
    "name": "User",
  },
}
`;

exports[`converters should resolve user role parameter 1`] = `
TestUserRole {
  "fallbackRoles": Array [],
  "id": "test0",
  "index": 0,
  "name": "Test1",
  "permissions": Array [],
  "type": EntityType {
    "name": "User",
  },
}
`;

exports[`converters should throw an error if member permission parameter is invalid 1`] = `
"[param: Invalid permission. Valid values: Test1, Test2, Test3]
Usage: /test {param}"
`;

exports[`converters should throw an error if member role parameter is invalid 1`] = `
"[param: Invalid role. Valid values: Test1, Test2]
Usage: /test {param}"
`;

exports[`converters should throw an error if user permission parameter is invalid 1`] = `
"[param: Invalid permission. Valid values: Test1, Test2, Test3]
Usage: /test {param}"
`;

exports[`converters should throw an error if user role parameter is invalid 1`] = `
"[param: Invalid role. Valid values: Test1, Test2]
Usage: /test {param}"
`;
